src/display/canvas.js

```diff
@@ -2973,19 +2977,25 @@ class CanvasGraphics {
   }
 
   beginAnnotations() {
-    this.save();
-    if (this.baseTransform) {
-      this.ctx.setTransform.apply(this.ctx, this.baseTransform);
-    }
+    // The annotations are drawn just after the page content.
+    // The page content drawing can potentially have set a transform,
+    // a clipping path, whatever...
+    // So in order to have something clean, we restore the initial state.
+    this.#restoreInitialState();
+    resetCtxToDefault(this.ctx, this.foregroundColor);
+    this.ctx.save();
   }
 
-  endAnnotations() {
-    this.restore();
-  }
+  endAnnotations() {}
 
   beginAnnotation(id, rect, transform, matrix, hasOwnCanvas) {
+    this.ctx.save();
     this.save();
```

Comment:
This `this.save`-call used to be "balanced" by `this.restore`-call in the old `endAnnotation`-method below, [note this code](https://github.com/mozilla/pdf.js/pull/14998/files#diff-76166cff75cae0ec9c490ae47aebbdadff02dbc1e26caaf93779a1bace64dc47L3053).

Unless I'm missing something that's no longer the case, and I thus wonder if this could possibly become a problem?

I can obviously see that there's a `this.restore`-call in the `#restoreInitialState`-method, but that one doesn't run unconditionally.

Reply:
`this.save` pushes a state in `stateStack` and `#restoreInitialState` restore until `stateStack` is empty so it should do the job.
So `save` is balanced with a `#restoreInitialState`.